File: dsa-problems/DP/solutions/javascript/DP-001-staircase-cracked-steps-maxjump.js
const readline = require("readline");
const MOD = 1000000007n;

class Solution {
  countWays(n, J, cracked) {
    if (cracked[n]) return 0;

    const dp = new Array(n + 1).fill(0n);
    dp[0] = 1n;
    let windowSum = 1n;

    for (let i = 1; i <= n; i++) {
      dp[i] = cracked[i] ? 0n : windowSum;
      windowSum = (windowSum + dp[i]) % MOD;

      const out = i - J;
      if (out >= 0) {
        windowSum = (windowSum - dp[out]) % MOD;
        if (windowSum < 0n) windowSum += MOD;
      }
    }
    return Number(dp[n] % MOD);
  }
}

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
const lines = [];
rl.on("line", (line) => lines.push(line.trim()));
rl.on("close", () => {
  let idx = 0;
  const [nStr, jStr] = lines[idx++].split(" ");
  const n = Number(nStr);
  const J = Number(jStr);
  const m = Number(lines[idx++]);

  const cracked = new Array(n + 1).fill(false);
  if (m > 0) {
    const arr = (lines[idx++] ?? "").split(" ").filter(Boolean).map(Number);
    for (const x of arr) {
      if (1 <= x && x <= n) cracked[x] = true;
    }
  }

  const sol = new Solution();
  console.log(sol.countWays(n, J, cracked));
});

File: dsa-problems/DP/solutions/cpp/DP-001-staircase-cracked-steps-maxjump.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    static const int MOD = 1000000007;

    int countWays(int n, int J, const vector<bool>& cracked) {
        if (cracked[n]) return 0;

        vector<long long> dp(n + 1, 0);
        dp[0] = 1;
        long long windowSum = 1;

        for (int i = 1; i <= n; i++) {
            dp[i] = cracked[i] ? 0 : windowSum;
            windowSum = (windowSum + dp[i]) % MOD;
            int out = i - J;
            if (out >= 0) {
                windowSum = (windowSum - dp[out]) % MOD;
                if (windowSum < 0) windowSum += MOD;
            }
        }
        return (int)(dp[n] % MOD);
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, J;
    cin >> n >> J;
    int m;
    cin >> m;
    vector<bool> cracked(n + 1, false);
    for (int i = 0; i < m; i++) {
        int x;
        cin >> x;
        if (1 <= x && x <= n) cracked[x] = true;
    }

    Solution sol;
    cout << sol.countWays(n, J, cracked) << "\n";
    return 0;
}
